packages/react-components/react-dialog/src/components/DialogSurface/useDialogSurface.ts

```diff
@@ -59,27 +58,30 @@ export const useDialogSurface_unstable = (
     }
   });
 
+  const backdropSlot = slot({
+    shorthand: backdrop,
+    elementType: 'div',
+    required: open && modalType !== 'non-modal',
+    defaultProps: { 'aria-hidden': 'true' },
+  });
+  if (backdropSlot) {
+    backdropSlot.onClick = handledBackdropClick;
+  }
+
   return {
-    components: {
-      backdrop: 'div',
-      root: 'div',
-    },
-    backdrop: resolveShorthand(backdrop, {
-      required: open && modalType !== 'non-modal',
-      defaultProps: {
-        'aria-hidden': 'true',
-        onClick: handledBackdropClick,
+    backdrop: backdropSlot,
+    root: rootSlot<DialogSurfaceSlots>({
```

Comment:
Looking on it now, I am thinking may be we can be more explicit. Here is an example:

```ts
const state = {
  root: slot.root(),
  backdrop: slot.optional(),
  foo: slot.required()
}
```

The downside that I see: it will be harder to have conditions for `required`/`optinal` as you have in this component.